sdk/core/core/src/az_json_builder.c

```diff
@@ -2,220 +2,627 @@
 
-    // check if the character has to be escaped.
+    switch (ch)
     {
-      az_span const esc = _az_json_esc_encode(c);
-      int32_t escaped_length = az_span_size(esc);
-      if (escaped_length)
+      case '\\':
+      case '"':
+      case '\b':
+      case '\f':
+      case '\n':
+      case '\r':
+      case '\t':
       {
-        AZ_RETURN_IF_NOT_ENOUGH_SIZE(remaining_json, escaped_length);
-        remaining_json = az_span_copy(remaining_json, esc);
-        self->_internal.length += escaped_length;
-        continue;
+        escaped_length += 2; // Use the two-character sequence escape for these.
+        break;
       }
+      default:
+      {
+        // Check if the character has to be escaped as a UNICODE escape sequence.
+        if (ch < 0x20)
+        {
+          escaped_length += 6;
+        }
+        else
+        {
+          escaped_length++; // No escaping required.
+        }
+        break;
+      }
+    }
+
+    i++;
+
+    // If this is the first time that we found a character that needs to be escaped,
+    // set out_index_of_first_escaped_char to the corresponding index.
+    // If escaped_length == i, then we haven't found a character that needs to be escaped yet.
+    if (escaped_length != i && *out_index_of_first_escaped_char == -1)
+    {
+      *out_index_of_first_escaped_char = i - 1;
+    }
+
+    // If the length overflows, in case the precondition is not honored, stop processing and break
+    // The caller will return AZ_ERROR_INSUFFICIENT_SPAN_SIZE since az_span can't contain it.
+    // TODO: Consider removing this if it is too costly.
+    if (escaped_length < 0)
+    {
+      escaped_length = INT32_MAX;
+      break;
     }
-    // check if the character has to be escaped as a UNICODE escape sequence.
-    if (c < 0x20)
+  }
+
+  // In most common cases, escaped_length will equal value_size and out_index_of_first_escaped_char
+  // will be -1.
+  return escaped_length;
+}
+
+static AZ_NODISCARD az_span _az_json_builder_escape_and_copy(az_span destination, az_span source)
+{
+  _az_PRECONDITION_VALID_SPAN(source, 1, false);
+
+  int32_t src_size = az_span_size(source);
+  _az_PRECONDITION(src_size <= _az_MAX_UNESCAPED_STRING_SIZE);
+  _az_PRECONDITION_VALID_SPAN(destination, src_size + 1, false);
+
+  int32_t i = 0;
+  uint8_t* value_ptr = az_span_ptr(source);
+
+  az_span remaining_destination = destination;
+
+  while (i < src_size)
+  {
+    uint8_t const ch = value_ptr[i];
+
+    switch (ch)
     {
-      uint8_t array[6] = {
-        '\\',
-        'u',
-        '0',
-        '0',
-        _az_number_to_upper_hex((uint8_t)(c / 16)),
-        _az_number_to_upper_hex((uint8_t)(c % 16)),
-      };
-
-      AZ_RETURN_IF_NOT_ENOUGH_SIZE(remaining_json, (int32_t)sizeof(array));
-      remaining_json = az_span_copy(remaining_json, AZ_SPAN_FROM_BUFFER(array));
-      self->_internal.length += (int32_t)sizeof(array);
-      continue;
+      case '\\':
+      case '"':
```

Comment:
To reduce code size consider a uint8_t variable that is set in each case to the char to append and after the switch, call az_span_copy_u8 once for the '\\' and once for the uint8_t variable. 

You could optimize even more by creating a 2-byte buffer with the first byte being '\\' and the second byte being the uint8_t char. Then you can call az_span_copy just once passing in a span over the 2-byte buffer.

Reply:
> To reduce code size consider a uint8_t variable that is set in each case to the char to append and after the switch, call az_span_copy_u8 once for the '\' and once for the uint8_t variable.

Great idea.

> You could optimize even more by creating a 2-byte buffer with the first byte being '\' and the second byte being the uint8_t char. Then you can call az_span_copy just once passing in a span over the 2-byte buffer.

I thought about this, but I believe our `copy` function for some threshold N performs worst than doing N `copy_u8` calls, and 2-4 might be that threshold. I will defer this particular change for later until after some investigation.